### ISSUE TITLE
Use CMake IMPORTED targets

### ISSUE THREAD

**Issue by gix** [conan-io/conan#168]
The current CMake integration seems to just create variables to the directories (include, lib, bin, …) for each dependency. For quite some time CMake has supported what they call imported targets. These group together properties. I would have expected conan to declare proper imported targets for each dependency. For example:

```
add_library(foo-1.2.3 SHARED IMPORTED)
set_target_properties(foo-1.2.3 PROPERTIES INTERFACE_INCLUDE_DIRECTORIES .../foo-1.2.3/include)
set_target_properties(foo-1.2.3 PROPERTIES IMPORTED_IMPLIB .../foo-1.2.3/lib/foo.lib)
set_target_properties(foo-1.2.3 PROPERTIES IMPORTED_LOCATION .../foo-1.2.3/bin/foo.dll)

add_library(bar-2.4.6 SHARED IMPORTED)
set_target_properties(bar-2.4.6 PROPERTIES INTERFACE_INCLUDE_DIRECTORIES .../bar-2.4.6/include)
set_target_properties(bar-2.4.6 PROPERTIES IMPORTED_IMPLIB .../bar-2.4.6/lib/foo.lib)
set_target_properties(bar-2.4.6 PROPERTIES IMPORTED_LOCATION .../bar-2.4.6/lib/foo.lib)
set_target_properties(bar-2.4.6 PROPERTIES INTERFACE_LINK_LIBRARIES foo-1.2.3)
```

This neatly captures the dependency graph and allows easily linking with only selected dependencies using `target_link_libraries(mytarget foo)` instead of setting several properties. Ideally this should also support multi-config CMake generators by specifying the paths for different build configurations (Debug, Release).

**Comment by memsharded** [comment]
I agree this is an interesting feature to have, but I would like to maintain the variables as they are too, because there is a reason for it. Defining only imported targets typically works well, but not always. Sometimes transitive properties accumulate in a non desired way, and difficult to debug and fix (I am thinking, for example in c++11, c++14 flags, with cmakes < 3.2). So having the variables is more verbose, but allows users to easily use what they want. They have full control in their cmake files of what flags, paths, etc. they decide to use and in which order.

Furthermore, most users have reported that they don't want to make big changes to their CMakeLists.txt, which are typically based on FindCMake stuff, and they want it to work the FindCMake way. So that is why we added the possibility that packages contribute their own FindCMakes, and add the package paths to CMAKE_MODULE_PATH

But I totally agree it would be great to have the IMPORTED targets available as well (maybe as opt-in, we dont want the lib names to collide with FindCMake stuff already working). 

I would like to code this, however, I see a small problem to implement it: the SHARED, INTERFACE type. Currently, there is no way, no standard property in conan that allows to know this, as it is up to the package creator. It is typical that package creators use some option of "shared" option, we could try to use it, but there is no way at all to detect the INTERFACE (e.g. header only) type. This is very cmake-ish, and conan is build system independent, so users have to define their package properties in their "conanfiles.py". Forcing a package developer that is using Makefiles as build system to add an "interface" attribute to their "conanfile.py", because CMake needs to know it (for consumers of the package) seems a bit weird. Would the UNKNOWN type work?

So my proposal here would be to try to add the IMPORTED targets to the generated cmake with UNKNOWN type, using the defined variables and see if it works. Any cmake experts out there willing to contribute this? I have my agenda full at least for 3-4 weeks :( 

Thanks very much for the suggestion!

**Comment by gix** [comment]
UNKNOWN changes how CMake links a target. It expects to find the import library in the IMPORTED_LOCATION property and skips IMPORTED_IMPLIB. There's another important feature which might not be immediately obvious from the snippet I posted. It also adds paths to the binaries which, at least with MSVC, are not required when linking. This has the huge benefit of allowing to gather all used libraries and copy them to the output directory or package them up when deploying.
This begs the question whether conan actually has enough information to implement this request.

And why is there no standardized conan property to distinguish shared and static packages? This seems like a huge oversight (especially seeing as licenses like LGPL require dynamic linking to avoid infection).

**Comment by memsharded** [comment]
Some other things that can be improved in CMake generation:

- Use per-target parameters when you can, the -m32 thing should be shipped with the dependency target
- Use target_compile_options, set target_properties, instead of the set(CMAKE_*_FLAGS)

**Comment by Lectem** [comment]
A few comments on what I saw above : 

> Sometimes transitive properties accumulate in a non desired way

indeed but

> and difficult to debug and fix (I am thinking, for example in c++11, c++14 flags, with cmakes < 3.2).

is *not* true. Using the properties [C_STANDARD](https://cmake.org/cmake/help/v3.1/prop_tgt/C_STANDARD.html) and [CXX_STANDARD](https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_STANDARD.html) should help a lot, and should work better unless the compiler isn't recognized.

As a side note, and this is mostly for things such as include dependencies, you can use the fact that you can specify the propagation of some properties using PRIVATE/PUBLIC/INTERFACE (see https://cmake.org/cmake/help/v3.0/manual/cmake-buildsystem.7.html#transitive-usage-requirements).

> we dont want the lib names to collide with FindCMake stuff already working

You could use namespaces the way FindQt4 and FindZLIB do it, resulting in targets named conan::THELIBTARGET.

> but there is no way at all to detect the INTERFACE (e.g. header only) type.

I don't know how conan works, but wouldn't checking the generated CONAN_LIBS_XXX against empty string be enough to know if it should be a header-only library ? Or perhaps I misunderstood the problem.

**Comment by memsharded** [comment]
Thanks, those suggestions are very useful. Just a question, we try to maintain compatibility on our side down to cmake 2.8.12, which (unfortunately) is still widely used as it is the one provided by some major distros. In fact the test suite should pass with 2.8.12. I recall that C_STANDARD and CXX_STANDARD where introduced in cmake 3.1, is that true?

Also, INTERFACE was introduced (at least for ``add_library()``) in cmake 3.0, this is the type of things we must be careful about. I really wish that all people upgraded to latest CMake, but the reality is a bit different, so it better be considered.

About the header only, thats a very good idea indeed, checking CONAN_LIBS should be enough, great.

**Comment by Lectem** [comment]
Compatibility is indeed an issue... Ubuntu and their cmake 2.8... But since it would be an addition to the current system, I am unsure if checking CMAKE_*_VERSION  would do it, as I don't know if the parser will still error out on unknown functions if hidden behind if/else ?
I suppose conan could check the cmake version installed on the system though.

**Comment by memsharded** [comment]
This issue has been practically duplicated by #601, so I think it is better to close this one and follow conversation and track progress in that one. Thanks!

**Comment by memsharded** [comment]
A solution using CMake targets for modern CMake has been implemented in #726, feedback from https://github.com/conan-io/conan/issues/601. Currently in develop branch, will be released in 0.17.
